crux-core/src/crux/query_state.clj

```diff
@@ -1,28 +1,58 @@
 (ns ^:no-doc crux.query-state
-  (:import (crux.api QueryState QueryState$QueryStatus QueryState$QueryError)))
+  (:import (crux.api IQueryState IQueryState$IQueryError IQueryState$QueryStatus)
+           (java.io Writer)))
+
+(defn ->query-status [status]
+  (case status
+    :failed IQueryState$QueryStatus/FAILED
+    :completed IQueryState$QueryStatus/COMPLETED
+    :in-progress IQueryState$QueryStatus/IN_PROGRESS))
+
+(defrecord QueryState$QueryError [type message]
+  IQueryState$IQueryError
+  (getErrorClass [this] type)
+  (getErrorMessage [this] type))
+
+(defrecord QueryState [query-id started-at finished-at status query error]
+           IQueryState
+           (getQueryId [this] query-id)
+           (getStartedAt [this] started-at)
+           (getFinishedAt [this] finished-at)
+           (getStatus [this] (->query-status status))
+           (getQuery [this] query)
+           (getError [this] error))
+
+(defmethod print-method QueryState [qs ^Writer w]
+  (.write w "#crux/query-state ")
+  (print-method (into {} qs) w))
+
+(defmethod print-method QueryState$QueryError [qs ^Writer w]
+  (.write w "#crux/query-error ")
+  (print-method (into {} qs) w))
 
 (defn <-QueryState [^QueryState query-state]
-  {:status (case (str (.status query-state))
-             "FAILED" :failed
-             "COMPLETED" :completed
-             "IN_PROGRESS" :in-progress)
-   :query-id (.queryId query-state)
-   :query (.query query-state)
-   :started-at (.startedAt query-state)
-   :finished-at (.finishedAt query-state)
-   :error (when-let [error (.error query-state)]
-            {:type (.errorClass error)
-             :message (.errorMessage error)})})
+      {:status      (case (str (.getStatus query-state))
```

Comment:
indentation here:
* `defn`/`case`/`when-let` are two-space
* we don't line up KVs of a map in Crux because it adds noise to diffs (the wider Clojure community is split) - just one space between K and V.

Think there's settings in IntelliJ for these - can pair if it'd be helpful